[Lester Ong] iP

lesterong opened the pull request:
## John
### Overview
John is the most generic command line-based chat bot to help you manage your tasks. John is
- Simple to use
- Fast
- Works everywhere, on Mac, Windows and Linux
- Free

### Getting Started
1. Download `jduke.jar` from [here](https://github.com/lesterong/ip/releases/download/A-Jar/jduke.jar).
2. Open up your Terminal, and navigate to the directory `jduke.jar` is in. 
3. Run `jduke.jar` by entering `java -jar jduke.jar` in the terminal.

### Features
- [x] list
- [x] add event
- [x] add deadline
- [x] add todo
- [x] add delete
- [x] find keywords
- [ ] renaming from jduke to john
- [ ] and more coming soon!

### For Java Programmers
Every time `john` is run, it calls the `main` method in `John.java`.
```java
public static void main(String[] args) {
  new John("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

lesterong committed:
Add greet, echo and exit

lesterong committed:
Add to list functionality

lesterong committed:
Add ability to mark and unmark tasks as done

lesterong committed:
Refactor into Task class

lesterong committed:
Add ToDo, Deadline and Event classes

lesterong committed:
Refactor while loop

lesterong committed:
Refactor into switch statements

lesterong committed:
Add support for todos, events and deadlines

lesterong committed:
Add semi-automated textui testing

lesterong committed:
Add DukeException class

lesterong committed:
Add errors handling for invalid user inputs

lesterong committed:
Add tests for exceptions

lesterong committed:
Change array to arraylist

lesterong committed:
Add support for deleting tasks

lesterong committed:
Handle commands using enum

lesterong committed:
Rename Duke to Jduke

lesterong committed:
Fix exception handling for unknown commands

lesterong committed:
Improve error handling for adding tasks

lesterong committed:
Refactor handling of strings

Refactor task details

Refactor handling of strings

Add test cases to check exception handling

Improve string handling

Simplify error handling

Use enums for commands instead

lesterong committed:
Ignore data directory

lesterong committed:
Add ability to save and update files locally

lesterong committed:
Handle date and time in event and deadline

lesterong committed:
Add ability to list task by date

lesterong committed:
Merge branch 'branch-Level-7'

lesterong committed:
Merge branch 'branch-Level-8'

lesterong committed:
Fix bug where delete removes multiple items

lesterong committed:
Add test cases for saving to storage

lesterong committed:
Refactor code into TaskList, Ui, Storage and Parser classes

lesterong committed:
Ignore storage data

lesterong committed:
Refactor into separate packages

lesterong committed:
Refactor enums into command classes

lesterong committed:
Merge remote-tracking branch 'remotes/upstream/add-gradle-support' into gradle

lesterong committed:
Update build.gradle to match directory

lesterong committed:
Refactor code for better OOP

lesterong committed:
Add unit tests for ui, task and commands

lesterong committed:
Add JavaDocs to public classes and methods

lesterong committed:
Add find command

lesterong committed:
Refactor to follow coding standards

lesterong committed:
Merge branch 'branch-A-JavaDoc'

lesterong committed:
Merge branch 'branch-A-CodingStandard'

lesterong committed:
Merge branch 'branch-Level-9'

lesterong committed:
Update deadline, event and list regex

lesterong committed:
Add checkstyle

lesterong committed:
Fix style violations

lesterong committed:
Rename jduke to john

lesterong committed:
Add GUI to John

- Fix horizontal scroll

- Refactor main

- Refactoring for oop

lesterong committed:
Add tests for GUI

lesterong committed:
Merge branch 'branch-Level-10'

lesterong committed:
Add varargs

lesterong committed:
Merge branch 'branch-A-Varargs'

LJXSean commented on src/main/java/john/John.java:
```diff
@@ -0,0 +1,74 @@
+ * The main logic of the John chatbot.
+ */
+public class John extends Application {
+    private TaskList tasklist;
```
Perhaps you could use camelCase here for variable names?

LJXSean commented on src/main/java/john/commands/IncorrectCommand.java:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Constructor for an IncorrectCommand.
+     * @param command The unknown command entered.
```
Should there be a blank line in between the description and parameter section?

LJXSean commented on src/main/java/john/commands/IncorrectCommand.java:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Constructor for an IncorrectCommand.
+     * @param command The known command entered.
```
Noticed the same issue here as well as a few other places

LJXSean commented on src/main/java/john/parser/Parser.java:
```diff
@@ -0,0 +1,125 @@
+        default:
+            return new IncorrectCommand(input);
+        }
+    }
```
Perhaps you could leave a blank line between methods to improve readability?

lesterong replied:
Hey @LJXSean, good catches, I've gone ahead to fix them!

LJXSean reviewed (commented):
LGTM almost, just a few nits to fix

lesterong committed:
Fix style violations

Polygonalr commented on src/main/java/john/task/Event.java:
```diff
@@ -0,0 +1,70 @@
+     * @return A string representing the storage format of the event.
+     */
+    @Override
+    public String toStorageFormat() {
+        return String.format(
+                "E | %d | %s | %s%s", (this.isCompleted ? 1 : 0), this.description,
+                this.date.format(Task.FORMATTER_INPUT_DATE), (this.time == null
+                        ? ""
+                        : " " + this.time.format(Task.FORMATTER_INPUT_TIME)));
+    }
```
This seems like a really [complicated expression](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-complicated-expressions). Do you think you can split it up into smaller steps to make the expression more legible? This applies for the other `toStorageFormat()` and `toString()` methods.

lesterong replied:
Good idea! I'll fix this in my next commit. 👍🏼

Polygonalr commented on src/main/java/john/data/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+     */
+    public String[] listTasks(String params) {
+        String[] tasksToShow = new String[tasks.size()];
+        if (tasks.size() == 0) {
+            return null;
```
Do you think it would be better if you try to make the [Happy Path more prominent](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent) by clustering this function's both `return null;` statements together? This also applies to the `findTasks` method.

lesterong replied:
This is an interesting suggestion, and I do agree that code readability may be improved with that. I'll probably need to put some thought into how to rearrange the flow of the program for this to work.

Polygonalr reviewed (commented):
Overall, code quality is good, aside from minor things I've pointed out. Variable and method names are proper, and all the methods are of acceptable length (<30 lines).

seanflyyy commented on src/main/java/john/parser/Parser.java:
```diff
@@ -0,0 +1,135 @@
+        String mainCommand = input.split(" ", 2)[0].toLowerCase();
+        String params = parseParams(input);
+
+        switch (mainCommand) {
```
I like the usage of the switch and case statement! Makes the code really readable

seanflyyy commented on src/main/java/john/ui/DialogBox.java:
```diff
@@ -0,0 +1,67 @@
+        return new DialogBox(text, img);
+    }
+
+    public static DialogBox getJohnDialog(String text, Image img) {
```
I like this method name!

seanflyyy commented on src/main/java/john/ui/MainWindow.java:
```diff
@@ -0,0 +1,63 @@
+
+    private John john;
+
+    private Image userImage = new Image(Objects.requireNonNull(
+            this.getClass().getResourceAsStream("/images/user.png")));
+    private Image johnImage = new Image(Objects.requireNonNull(
+            this.getClass().getResourceAsStream("/images/john.png")));
```
You can consider setting this to final since it is unlikely the name should change

seanflyyy commented on src/test/java/john/commands/FindCommandTest.java:
```diff
@@ -0,0 +1,29 @@
+
+public class FindCommandTest {
+    @Test
+    public void executeTestNoResults() {
```
Perhaps you can consider setting the test methods in the following format?
https://se-education.org/guides/conventions/java/intermediate.html

![image](https://user-images.githubusercontent.com/50956270/188148781-7ee9f1e3-5a34-4eb7-896d-0619e57c2ef0.png)

lesterong replied:
I'm working on this, thanks for the suggestion!

seanflyyy commented on src/main/java/john/ui/Ui.java:
```diff
@@ -0,0 +1,138 @@
+     * @param params The parameters of the query.
+     * @return A string showing that there are no tasks for the query.
+     */
+    public String showNoTasks(TaskList taskList, String params) {
```
I like your method naming

seanflyyy reviewed (commented):
Overall really clean code. Could hardly find any issues. Just a few minor things that you can consider!

lesterong committed:
Improve code quality

The happy path is not obvious in the TaskList class, the toString
and toStorageFormat in Deadline, Event, and Todo classes are long
and complex.

These above-mentioned issues reduce the code quality and make the
code harder to read, and thus difficult to maintain.

By extracting methods into variables in the Task class and its
subclasses. Shifted the guard clause in the TaskList, and indented
the edge cases.

These changes make the code more maintainable and readable, as
one can see the edge cases and variables in one glance easily.

lesterong committed:
Add assertions to commands and tasklist classes

There are no assertions within the code across all classes.

There is a higher risk of bugs when there are no assertions,
since we are not verifying necessary conditions.

Adding assertions to the Command allows the code to terminate
once a bug is detected during an assertion failure, reducing
the risk of bugs, for example, when the input position into
the mark command is smaller than 0, there is most likely a
bug in the code.

lesterong committed:
Use streams in TaskList

lesterong committed:
Merge pull request #1 from lesterong/branch-A-Assertions

Add assertions to commands and tasklist classes

lesterong committed:
Merge branch 'master' into branch-A-CodeQuality

lesterong committed:
Merge pull request #2 from lesterong/branch-A-CodeQuality

Improve code quality

lesterong committed:
Merge branch 'master' into branch-A-Streams

lesterong committed:
Merge pull request #3 from lesterong/branch-A-Streams

Use streams in TaskList

lesterong committed:
Add CI

lesterong committed:
Fix line separators for windows

lesterong committed:
Add basic help command

lesterong committed:
Remove extra new line

lesterong committed:
Add tests for HelpCommand, TaskList and Parser classes

lesterong committed:
Add more error handling

lesterong committed:
Add dummy commit to force CI

lesterong committed:
Initial styling

lesterong committed:
Improve the GUI

lesterong committed:
Update help message

lesterong committed:
Add user guide

lesterong committed:
Remove failing test

lesterong committed:
Fix issue with github pages

lesterong committed:
Update Ui image

lesterong committed:
Style user guide

lesterong committed:
Fix styling for windows

lesterong committed:
Fix issue with initial focus
